### ISSUE TITLE
NavigationExtras when returning URLTree from guard

### ISSUE THREAD

**Issue by Airblader** [angular/angular#27148]
# 🚀 feature request

### Relevant Package
This feature request is for @angular/router

### Description
In #26521 (Angular 7.1.0), a feature was introduced that allows returning a `URLTree` from guards in order to initiate a redirect. However, this doesn't seem to allow using `NavigationExtras` such as `skipLocationChange`.

Since the change in #26521 wasn't just syntax sugar, but is relevant to have predictable behavior when dealing with multiple guards, it should be considered to somehow allow using these options when returning a `URLTree` as well.

### Describe the solution you'd like
I don't have a really good proposal at this moment apart from next to returning `URLTree` also returning some kind of

```
// TODO: Find a better name?
interface NavigationOptions {
  urlTree: URLTree;
  navigationExtras: NavigationExtras;
}
```

such that I could use

```
class AwesomeGuard implements CanActivate {
  constructor(private router: Router) {}

  canActivate(): NavigationOptions | Observable<NavigationOptions> {
    return {
      urlTree: this.router.parseUrl("…"),
      navigationExtras: {
        skipLocationChange: true,
      },
    };
  }
}
```

I guess the exact typing of the guard would have to be

```
// Just a placeholder name
type GuardReturnType = boolean | URLTree | NavigationOptions;

interface CanActivate {
  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): 
    GuardReturnType 
    | Promise<GuardReturnType> 
    | Observable<GuardReturnType>
}
```

Admittedly, the return type of this function is growing increasingly complex with this.

### Describe alternatives you've considered
Possible options that I don't think would make good solutions would be

* Returning an array `[URLTree, NavigationExtras]` (this just feels unclean)
* Extending URLTree with the navigationExtras and extend parseUrl to take them as well (this muddies the entire API for a change in a specific location)

**Comment by Airblader** [comment]
I guess the exact typing of the guard would have to be

```
// Just a placeholder name
type GuardReturnType = boolean | URLTree | NavigationOptions;

interface CanActivate {
  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): GuardReturnType | Promise<GuardReturnType> | Observable<GuardReturnType>
}
```

**Comment by Airblader** [comment]
Actually, if going down this route, we'd probably need to split `NavigationExtras` into multiple parts as it also contains things such as  `queryParams`, which is already part of the `URLTree`. We'd need a base that only contains the extras like `skipLocationChange`. This should still be a compatible change, though, as it can be

```
// Again, names are placeholders

interface NavigationExtrasBase {
  skipLocationChange?: boolean
}

interface NavigationExtras extends NavigationExtrasBase {
  // … as today, just without skipLocationChange
}

interface NavigationOptions {
  urlTree: URLTree;
  // … and now this base type here:
  navigationExtras: NavigationExtrasBase;
}
```

**Comment by Airblader** [comment]
@jasonaden Thanks for the reply! Our concrete usecase is a 403/404 redirect where we want to show an error page but retain the URL.

**Comment by Airblader** [comment]
@jayoungers I think in your second example replaceUrl would also be a reasonable choice, so I think just setting skipLocationChange doesn't cut it here. Also that would be pretty unexpected and confusing in my opinion.

**Comment by jayoungers** [comment]
@Airblader I think replaceUrl would probably work in all of the scenarios.

When the `UrlTree` result for `canActivate` was added as an option, in my mind the functionality was now:

- Yes, you can go to this route (`true`)
- No, you can't go to this route (`false`)
- You're going to the wrong route - you should be requesting this route instead (`UrlTree`)

Regarding the scenario: is the route replacement the unexpected/confusing part? or just the fact the url is altered?  My scenario probably isn't the best representation, but if the issue is around the route being altered/standardized in general, I think that's a pretty common practice (news sites usually rewrite the url with the article title, stackoverflow will do this as well if the url only contains the question id, etc)

**Comment by Airblader** [comment]
> is the route replacement the unexpected/confusing part?

The unexpected part is that the navigation works differently than any other navigation where I don't set any navigation extras. IMHO here Angular would suddenly dictate opinion over what navigation does or doesn't make sense, and I would like to avoid that. Everywhere else the user gets to choose the options, so it should be the same here.

**Comment by jayoungers** [comment]
I see.  I think the answer comes down to what the intention of returning `UrlTree` is:

I'm looking at is as though the requested route is not valid under any circumstance.  If we're looking at it as a short hand for the navigate method, then what you're saying makes perfect sense.

In my mind I would have had the functionality setup so that if `UrlTree` is returned, then that is the route the user intended to go to (the guard result wasn't `true`, it wasn't `false`, it just lead to the correct path).  

In scenarios such as authentication for example, where the user isn't logged in or the session has expired, I don't think `UrlTree` is the correct response:  it should be `false` (this route is valid, but you don't have access to it), and the guard `router.navigate`s to the login page.

**Comment by Airblader** [comment]
But the benefit and entire point of allowing returning a tree is that it makes how multiple guards interact predictable and stable. This feature wasn't about specific usecases. I'd like to start returning URL tree in as many cases as I can (whenever I don't return true).

**Comment by jayoungers** [comment]
I agree: based on the original requirements I would say I'm definitely not looking at the feature as it was intended.

If you don't mind, just out of curiosity, would you happen to have an example scenario of multiple guards firing on a route that result in different `UrlTree`s?

**Comment by Airblader** [comment]
I don't have a concrete usecase for this from my own projects, but admittedly I don't use anywhere near the full feature set of the router. I think with its extensive support of multiple outlets and the like such scenarios are very likely, for example authenticating on different resources or something.

**Comment by axelboc** [comment]
Two concrete use cases I have throughout my project in `canActivate` guards:

#### 1) Redirecting to a 404 page when trying to access an unknown resource

- The expected behaviour is that of `skipLocationChange`: the user should see the URL of the unknown resource, so they can fix a typo in the URL and try again, for instance.
- With the `UrlTree` technique, at the end of the redirect, the URL shows `/404` and the user can navigate back.

#### 2) Performing a 301 redirect to a resource's canonical URL

- This is useful when a resource's URL can be renamed or aliased, for instance. The expected behaviour is that of `replaceUrl` : the user should see the canonical URL without being able to navigate back to the obsolete/aliased URL.
- With the `UrlTree` technique, the user can navigate back to the resource's obsolete/aliased URL.

**Comment by scott-ho** [comment]
> But the benefit and entire point of allowing returning a tree is that it makes how multiple guards interact predictable and stable. This feature wasn't about specific usecases. I'd like to start returning URL tree in as many cases as I can (whenever I don't return true).

@Airblader good point.

---

the current code is using a serialized url to redirecting, [see the source](https://github.com/angular/angular/blob/a28b3e3359/packages/router/src/router.ts#L744). 

@jasonaden Concrete use cases are not the major concern for the feature request. There are so many possibilities.

`would that cause problems in the application` that's an application problem but not a framework problem.

Since for the first glance, you don't have valid point about what will be bad for the framework. I believe we could move on to make this happen.

**Comment by jasonaden** [comment]
@Airblader @jayoungers @scott-ho @axelboc 

Thanks for all the comments on this issue. I want to get this resolved somehow in the very near future. Ideally in the `RC` period of version 8 (there might be a bit of an issue doing so because it would be adding a feature).

I'm happy to see the `404` and `301` examples above from @axelboc.

We were looking at an issue internally and found that, with the redirects being done the way they currently are, the `back` button is essentially broken. I was looking at making a fix such that returning a `UrlTree` will cause the same redirect, but will always be done with `replaceUrl`. This would be turned on by default, because without it hitting `back` will end up redirecting again, and the user will not be able to go `back` past the page they redirected to.

I would love some feedback from anyone using this feature or wanting to use this feature. We can get this implemented quickly and pushed out as part of `RC` as a `fix` rather than the full feature described above.

That being said, the original design did contain an option to provide `NavigationExtras` but we didn't have concrete use cases where it would be needed. Now, with those provided, we should be able to implement this API as well.

**Comment by jayoungers** [comment]
@jasonaden - If I'm following the proposed `v8` fix (if the guard results in a `UrlTree`, then replace the requested route with the `UrlTree` route), that's good with me:  from my earlier comments I think that's how I assumed the feature was intended to be used.

With just that fix, I believe that should resolve all of my scenarios, and the 2nd example @axelboc provided.  I think the 1st example would need the additional functionality

**Comment by axelboc** [comment]
@jasonaden the v8 fix you're suggesting sounds great. You're right, it would at least fix the back button for now.

**Comment by dubeg** [comment]
I was confused that this wasn't implemented because of the function createUrlTree on the router, which has the NavigationExtras as a parameter. I guess that wasn't intended to be used within a guard.

Just to add my own case, I had a CanActivate guard verifying that the user is part of a certain group before accessing a page. If he isn't, then redirect to an AccessDenied page that can display the group which the user lacks. The group name would've been passed via NavigationExtras' state within the guard, since the guard is better positioned to know the group name, and the AccessDenied component may potentially be used in different contexts, by different guards. 

Hope that makes sense.

**Comment by laurentgoudet** [comment]
> We were looking at an issue internally and found that, with the redirects being done the way they currently are, the back button is essentially broken. I was looking at making a fix such that returning a UrlTree will cause the same redirect, but will always be done with replaceUrl. This would be turned on by default, because without it hitting back will end up redirecting again, and the user will not be able to go back past the page they redirected to.

As this been fixed? We've just discovered that returning a `UrlTree` indeed breaks navigation history, and looking at the changelog it doesn't seem to have been fixed in v8 (nor v8.1 beta yet)?

**Comment by jasonaden** [comment]
My apologies on this. It should have been in 8, but got missed. See above for the fix. This should land shortly.

**Comment by pjc23** [comment]
Any plan to re-release this? 15e397816f98ec16839c30fd5c1ea01c7444fb84 got reverted in 0de5d79bf620ecf3c52f2c2c0a7c915aef6e2f34

**Comment by csvn** [comment]
Similar to other mentioned use cases, we would like to preserve the URL for error pages:

- **403**: Logged in, but is missing permission (should not occur by regular app usage). Manipulating URL or reloading page after permissions are revoked could trigger it. Show a generic `ForbiddenComponent` with error message (known by guard)
- **404**: For cases where a route matches (e.g. `/user/:id`), but a resource does not exist (or has been removed. A guard could be used to show a `NotFound` component

We also use a lot of redirects currently to populate query parameters. It would be great if it was possible to redirect and add missing parameters:

```ts
canActivate(route: ActivatedRouteSnapshot): boolean | UrlTree {
  if (route.params.page) return true;
  const queryParams = { page: '1' };
  const urlTree = this.router.createUrlTree(['.'], { relativeTo: route, queryParams });
  return { urlTree, preserveQueryParams: true, queryParamsHandling: 'merge' };
}
```

So basically, all options from [NavigationExtras](https://angular.io/api/router/NavigationExtras) except `relativeTo`, `queryParams`, and `fragment` (since these are handled by `router.createUrlTree`) would be great to have IMO. 😀

**Comment by FDIM** [comment]
Just stumbled upon this issue, I guess the only workaround now not break back button is to return false and schedule a navigation in set timeout in a guard. 

Interesting to see that not many people are hitting this issue

**Comment by ErikWitkowski** [comment]
> Just stumbled upon this issue, I guess the only workaround now not break back button is to return false and schedule a navigation in set timeout in a guard.
> 
> Interesting to see that not many people are hitting this issue
> 
> EDIT:
> This seems to work as well in canActivate:
> 
> ```
> this.router.events.pipe(
>   filter(ev => ev instanceof NavigationCancel),
>   take(1)
> ).subscribe(() => {
>   this.router.navigateByUrl(redirectUrl, { replaceUrl: true });
> });
> ``
> ```

Thanks for the workaround! But what if we want to preserve NavigationExtras? 

Let's say we want to use canActivate to redirect the user to other route but we want to keep NavigationExtras.

**Comment by FDIM** [comment]
@ErikWitkowski inside canActive method, you can get navigation object via `router.getCurrentNavigation()` - it will have extras property that includes original options passed via navigate method

**Comment by a88zach** [comment]
IMO, the developer defined state that gets added to the NavigationExtras should also be passed through when returning a UrlTree.  Otherwise this additional information gets lost during the redirect

**Comment by simeyla** [comment]
Please don't forget about some sort of [feature parity](https://github.com/angular/angular/issues/29089) with `Resolve<T>` when considering changes to URLTrees from guards.

Being able to redirect elegantly from a Resolver is very important - especially in the case where an invalid URL is provided (eg. a mistyped ID).

**Comment by sasoori** [comment]
Hey, wanted to know the status of this feature request. No update for few months now, was it missed again?

**Comment by troydietz** [comment]
We are using a route guard to redirect users to a loading page before our route resolvers run.  Without a loading page, if an authenticated user navigates directly to a url that requires a lot of data to be resolved, the user is presented with a blank white page with nothing indicating anything is working.  Not ideal.

To get around this, we have implemented a loading page.  We'd like the loading page url to never be seen by our users.  However, because the `skipLocationChange` is not working with `Router.createUrlTree`, we cannot achieve this.  Current behavior: User see's `/desired/path` then `/loading`, then `/desired/path` in their browser.  Ideally the url would not change and we could seamlessly present a loading page while also using route resolvers.

**Comment by jayoungers** [comment]
You may be better off with some sort of view cover indicator that exists at a root level to indicate the route is loading:  I think you'll run into issues eventually for more complicated scenarios with redirecting from the intended route and back

**Comment by FDIM** [comment]
In all the cases at work we handle fetching and loading indicators inside page components

Personally I do not understand the need of resolvers (now and in angularjs)

**Comment by pickfire** [comment]
We handle it using guards, redirect to another page if it hits the criteria.

**Comment by egorkarimov** [comment]
I had the same idea in mind when I was writing in my guard something like this:
`this.router.navigateToUrl(PAGE_403, {state: state});`.
Do you working on this feature? In which version of Angular it will be available?

**Comment by atscott** [comment]
Also related to #16981

**Comment by rhutchison** [comment]
The use case I have is similar to https://github.com/angular/angular/issues/17004 - being able to render a 404 response for the current route.  I don't want to redirect to a '404' / 'not-found' route.

My current workaround is slightly different than others.

```typescript
  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot) {
    // BUG: https://github.com/angular/angular/issues/27148
    // return this._router.createUrlTree(['not-found'], {
    //   skipLocationChange: true,
    // });

    return this.profileExists(route.params.id).pipe(
      mapTo(true),
      catchError(() =>
        this.store
          .dispatch(
            new Navigate(['not-found'], null, {
              skipLocationChange: true,
            })
          )
          .pipe(
            tap(() => this._location.go(state.url)),
            mapTo(false)
          )
      )
    );
  }
```

I found the other solutions were overwriting the previous state, instead of the intended state, which caused issue with back/forward.

**Comment by palendrance** [comment]
+1 for this feature request, I was honestly surprised that you can't utilize anything like navigationExtras in a guard when you can do it everywhere else.

I was hoping to build some process validation into my navigation guards to prevent users from getting clever and trying to bypass the usual workflow by changing the URL. I can still do that the way thing are now, but I can't funnel any custom data back through the navigationExtras in a neat and readable way.

I was initially thinking a good solution would be to provide some sort of "alternateRouterCommand" interface on the guard itself, but I think I might have a better solution that covers more scenarios. Would it be reasonable to provide an "onGuarded" callback property on a Route that aggregates the values returned by negative guards into an array until all guards have been processed, then calls the function you provide in the "onGuarded" property with the array of values returned from failed guards, target route, and current route? That would allow developers to build dynamic responses from rejected routes, predict the order in which guards will run, and put a lot of their error handling in a neat, centralized location.

**Comment by UrbanNuke** [comment]
Are any changes?

**Comment by timminss** [comment]
> Similar to other mentioned use cases, we would like to preserve the URL for error pages:
> 
> * **403**: Logged in, but is missing permission (should not occur by regular app usage). Manipulating URL or reloading page after permissions are revoked could trigger it. Show a generic `ForbiddenComponent` with error message (known by guard)
> * **404**: For cases where a route matches (e.g. `/user/:id`), but a resource does not exist (or has been removed. A guard could be used to show a `NotFound` component
> 
> We also use a lot of redirects currently to populate query parameters. It would be great if it was possible to redirect and add missing parameters:
> 
> ```ts
> canActivate(route: ActivatedRouteSnapshot): boolean | UrlTree {
>   if (route.params.page) return true;
>   const queryParams = { page: '1' };
>   const urlTree = this.router.createUrlTree(['.'], { relativeTo: route, queryParams });
>   return { urlTree, preserveQueryParams: true, queryParamsHandling: 'merge' };
> }
> ```
> 
> So basically, all options from [NavigationExtras](https://angular.io/api/router/NavigationExtras) except `relativeTo`, `queryParams`, and `fragment` (since these are handled by `router.createUrlTree`) would be great to have IMO. 😀

Did you manage to come up with a solution to this issue? I am trying to implement the same 403 use case you mentioned. Thanks!

**Comment by csvn** [comment]
@timminss not a good one, I'm afraid. We did something similar to the below code snippet, but using a generic "error" page for various errors instead of a specific route for each error. The errors generated are from the `HttpClient` when resolving data, so they had `err.status` that we could check for routing purposes.

This was the best we could come up with 2-3 years ago when we wrote the code. We weren't able to find much documentation for more advanced routing use-cases at the time.

```ts
@Injectable()
export class Route403Service {
  constructor(private router: Router, private location: Location) {}

  init() {
    let navError: NavigationError;
    this.router.events
      .pipe(filter((e): e is NavigationError => e instanceof NavigationError))
      .subscribe(e => navError = e);

    this.router.errorHandler = err => {
      if (err.status === 403) {
        this.router.navigate(['/not-found'], { skipLocationChange: true }).then(() => {
          // Show the route that couldn't be navigated to in the browser
          this.location.replaceState(navError.url);
        });
      }
    };
  }
}
```

**Comment by mabraithwaite** [comment]
@csvn is there a reason you used `this.location.replaceState()` over `this.location.go()` which would push the url onto the browser state. That way they can push the back button and go to the last place that was "working".

**Comment by KrakenTyio** [comment]
Is some new progress with that?

**Comment by atscott** [comment]
Hi, no progress or ETA on this, but here to provide some alternative workaround hackery for the `skipLocationChange` redirect:

TL;DR - for the `skipLocationChange` case, you can add this one-liner before the `return <urlTree>` in your guard:
 `this.router.getCurrentNavigation().extras.skipLocationChange = true;`
 Please add a test for this in with `RouterTestingModule`

Two pre-understanding points:
* When the `Router` uses the `UrlTree` for redirecting, it inherits the `skipLocationChange` value from the last request ([code reference](https://github.com/angular/angular/blob/6967f3ca28ab30bc9d984e10764e87ad40535bb5/packages/router/src/router.ts#L893). 
* The `currentNavigation` property in the `Router` gets the extras from the navigation request without doing a deep copy ([code reference](https://github.com/angular/angular/blob/6967f3ca28ab30bc9d984e10764e87ad40535bb5/packages/router/src/router.ts#L576))

Because of the above two points, you can actually simply do `this.router.getCurrentNavigation().extras.skipLocationChange = true;` before the `return <urlTree>` in your guard. Doing this certainly depends on the `Router` implementation details so please add a test for your guard using `RouterTestingModule` and understand that it is _possible_ that this could break with changes to the `Router` code (and those changes would not be considered breaking changes since the internal implementation isn't public API).

**Comment by BigPackie** [comment]
> @timminss not a good one, I'm afraid. We did something similar to the below code snippet, but using a generic "error" page for various errors instead of a specific route for each error. The errors generated are from the `HttpClient` when resolving data, so they had `err.status` that we could check for routing purposes.
> 
> This was the best we could come up with 2-3 years ago when we wrote the code. We weren't able to find much documentation for more advanced routing use-cases at the time.
> 
> ```ts
> @Injectable()
> export class Route403Service {
>   constructor(private router: Router, private location: Location) {}
> 
>   init() {
>     let navError: NavigationError;
>     this.router.events
>       .pipe(filter((e): e is NavigationError => e instanceof NavigationError))
>       .subscribe(e => navError = e);
> 
>     this.router.errorHandler = err => {
>       if (err.status === 403) {
>         this.router.navigate(['/not-found'], { skipLocationChange: true }).then(() => {
>           // Show the route that couldn't be navigated to in the browser
>           this.location.replaceState(navError.url);
>         });
>       }
>     };
>   }
> }
> ```

I would say this is almost good, because the `navError.url` is the whole url not just the path. Most likely you would like to replace only the **path part** of the url:

```ts
    const currentPath = this.router.getCurrentNavigation()?.extractedUrl.toString();
```

I needed to use it in an `HttpInterceptors` to handle errors - navigating user to a specific error page while keeping the original requested path in the url bar:

```ts
@Injectable()
export class ErrorResponseInterceptor implements HttpInterceptor {
  constructor(private router: Router, private location: Location) {}

  intercept(
    request: HttpRequest<unknown>,
    next: HttpHandler
  ): Observable<HttpEvent<unknown>> {
    return next.handle(request).pipe(
      // retry(1),
      catchError((error: HttpErrorResponse) => {
        return this.handleError(error);
      })
    );
  }

  handleError(error: HttpErrorResponse) {
    let currentPath = this.router
      .getCurrentNavigation()
      ?.extractedUrl.toString();
    if (error) {
      let targetPath;
      switch (error.status) {
        case 500:
          targetPath = '500';
          break;
        case 404:
          targetPath = '**';
          break;
        default:
          targetPath = '**';
          break;
      }

      this.router
        .navigateByUrl(targetPath, { skipLocationChange: true })
        .then(() => {
          if (currentPath) {
            this.location.replaceState(currentPath);
          }
        });

      return EMPTY;
    }

    return throwError(error);
  }
}
```

Of course routing must be defined somewhere, ideally in some eagerly loaded module , e.g. the app-routing.module.ts : 

```ts
  \\... some paths refined in the Routes array  and than this at the end
  { path: '500', component: P500Component },
  { path: '**', component: P404Component }, // Wildcard route for a 404 page
```

and interceptor registered in the providers array of the app.module.ts : 

```ts

  \\...
  \\ interceptor order matters !
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ErrorResponseInterceptor,
      multi: true,
    },

 \\...
```

**Comment by chjernig-msft** [comment]
+1 to this issue as having navigationExtras when creating a UrlTree allows canActivate to preserve extras like state if the url is changed.
